Add entity to bookcal

G-ladal opened the pull request:
# New prepare DB for bookcal compatibility with multicompany
Core database miss "entity" for bookcal tables, it will allow compatibility with MC (in following pushes).

G-ladal committed:
[New] prepare bookcal DB for multicompany

Add entity to calendar and availability tables

G-ladal committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into develop

G-ladal commented:
this is my 1st PR, I tried to be compliant, but please point me to any mistake (for next ones).

frederic34 commented:
@G-ladal do not forget to make alse changes in mysql/tables

G-ladal committed:
Fix propagate bookcal mods to main tables

add entity field for bookcal to install scripts (not only migration)

G-ladal committed:
Merge branch 'develop' into develop

G-ladal committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into develop

G-ladal commented:
@frederic34 Thanks ! added.

frederic34 commented:
and now you can wait for validation before next step

hregis commented:
Hello @G-ladal 

I don't know this module in all these details... if you create an element in "llx_bookcal_calendar" and you share it between several entities... the linked elements in the "llx_bookcal_availabilities" table will be unique or will you could have different elements in "llx_bookcal_availabilities" which will be linked with the same "llx_bookcal_calendar"?

if you don't understand anything I said and if you speak French... call me! ;-)

G-ladal commented:
@hregis : module isn't mine, but i'm using it (FYI it's 80% functional with MC in my offline repo).
=> i don't have your phone number (tried to find it in inoBox website)

i think i understood what you mean, let me pseudo-code/desing what can exist :
```
foreach(Entity in entities) {
    foreach(calendar in $entity.calendar_list) {
        foreach (availability in $calendar.availability_list) {
            print $availability.starttime . " to ". $availability.endtime;
        }
    }
}
```
with calendar as an agenda object

So :
we need do split both lvl (entity wise) for full granularity (and be able to hide/share cal/agenda AND avaibilities between entities).

eldy commented on htdocs/install/mysql/migration/19.0.0-20.0.0.sql:
```diff
@@ -292,3 +292,7 @@ ALTER TABLE llx_receptiondet_batch CHANGE COLUMN fk_commandefourndet fk_elementd
 ALTER TABLE llx_projet ADD COLUMN ref_ext varchar(50) after ref;
+
+-- bookcal
+ALTER TABLE llx_bookcal_availabilities ADD COLUMN entity integer DEFAULT 1 NOT NULL AFTER rowid
```
The bookcal_availabilities are linked to a bookcal_calendar (child of bookcal_calendar).
So no need for entity in this table. We only need it into bookcal_calendar.

G-ladal replied:
hi @eldy : there might be something I didn't understand in PHP, but I had to add entity field to availabilities table to be able to :
- see (and set) availability on a specific (shared) calendar (=> from ent2, set availability on a ent1 calendar)
- not have warning/error on missing "entity" field in availability_card.php :
 ```
$upload_dir = $conf->bookcal->multidir_output[isset($object->entity) ? $object->entity : 1].'/availabilities';
```
-> I tried without it (trying to modify the less code possible), but trial and errors forced me to do it.


(=>continuing php investigations in the meantime)

eldy replied:
Dev was initialized by @Hystepik.
I suggest to first submit a PR with the addition of the database structure and only database structure. So adding the new field entity into llx_bookcal_calendar.
Even if not used by code, validation of structure addition is first step to progress...

If a booking calendar is defined into entity X, all availabilities will be saved as linked to booking calendar (so in entity X). 
And if you decide to share the calendar between entities, X and Y, it means you want to share also the availabilities. So no need for entity field into the table of availability.

G-ladal replied:
![Screenshot 2024-04-04 200057](https://github.com/Dolibarr/dolibarr/assets/165890649/a9a9418d-c55a-412b-90bd-aea5e5356435)
![Screenshot 2024-04-04 195948](https://github.com/Dolibarr/dolibarr/assets/165890649/5e81e369-2d10-4fa2-8b8a-62577264a92f)


Last screen is from "List calendars", 1st one is from "create availability"

G-ladal committed:
add entity to bookcal_calendar only

as per requested, add entity to bookcal_calendar only

	modified:   htdocs/install/mysql/migration/19.0.0-20.0.0.sql
	modified:   htdocs/install/mysql/tables/llx_bookcal_calendar-bookcal.sql

G-ladal committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into develop

G-ladal committed:
Merge branch 'Dolibarr:develop' into develop

G-ladal commented:
hopefully PR is ok now

Sorry i'm not used to git (i'm old school diff/patch addict), there's probably tons of useless commit/push/pull/reset/merge.

G-ladal committed:
Revert changes on availabilities table

eldy commented:
> hopefully PR is ok now
> 
> Sorry i'm not used to git (i'm old school diff/patch addict), there's probably tons of useless commit/push/pull/reset/merge.

Don't worry. Your PR has been merged using the "squash" mode so all your commit appears as 1 into the official repository after the merge.
